rust/lance-index/src/vector/hnsw/builder.rs

```diff
@@ -183,92 +197,155 @@ impl HNSWBuilder {
+            select_neighbors_heuristic(
+                &cur_level,
+                query,
+                &candidates,
+                self.params.m,
+                self.params.extend_candidates,
+            )
+            .collect()
+        } else {
+            select_neighbors(&candidates, self.params.m)
+                .cloned()
+                .collect()
+        };
+
+        Ok((candidates, neighbors))
     }
 
-    /// Build the graph, with the provided [VectorStorage] as backing storage for HNSW graph.
-    pub fn build_with(&mut self, storage: Arc<dyn VectorStorage>) -> Result<HNSW> {
+    fn connect(&mut self, u: u32, v: u32, dist: OrderedFloat, level: u16) {
+        self.nodes[u as usize].add_neighbor(v, dist, level);
+        self.nodes[v as usize].add_neighbor(u, dist, level);
+    }
+
+    fn prune(&mut self, node: u32, level: u16) {
+        let level_neighbors = &self.nodes[node as usize].level_neighbors[level as usize];
+        if level_neighbors.len() <= self.params.m_max {
+            return;
+        }
+
+        let level_view = BuilderLevelView::new(level, self);
+        let neighbors: Vec<OrderedNode> = level_neighbors.iter().cloned().collect();
+
+        let new_neighbors = select_neighbors_heuristic(
+            &level_view,
+            self.vectors.vector(node),
+            &neighbors,
+            self.params.m_max,
+            self.params.extend_candidates,
+        )
+        .collect();
+
+        self.nodes[node as usize].level_neighbors[level as usize] = new_neighbors;
+    }
+
+    /// Build the graph, with the already provided `VectorStorage` as backing storage for HNSW graph.
+    pub fn build(&mut self) -> Result<HNSW> {
         log::info!(
             "Building HNSW graph: metric_type={}, max_levels={}, m_max={}, ef_construction={}",
             self.vectors.metric_type(),
             self.params.max_level,
             self.params.m_max,
             self.params.ef_construction
         );
-        for _ in 0..self.params.max_level {
-            let mut level = GraphBuilder::new(self.vectors.clone());
-            level.insert(0);
-            self.levels.push(level);
-        }
+
+        self.nodes
+            .push(GraphBuilderNode::new(0, self.params.max_level as usize));
 
         for i in 1..self.vectors.len() {
             self.insert(i as u32)?;
         }
 
-        if log_enabled!(Info) {
-            for (i, level) in self.levels.iter().enumerate() {
-                info!("HNSW level {}: {:#?}", i, level.stats());
-            }
-        }
-
-        let graphs = self
-            .levels
-            .iter()
-            .map(|l| HnswLevel::from_builder(l, storage.clone()))
-            .collect::<Result<Vec<_>>>()?;
         Ok(HNSW::from_builder(
-            graphs,
+            self,
             self.entry_point,
             self.vectors.metric_type(),
             self.params.use_select_heuristic,
         ))
     }
 }
+
+pub struct BuilderLevelView<'a> {
```

Comment:
Could you add a comment about what is this. what is the logical relationship between this and HnsfBuilder.

Also should this be `private` or `pub(crate)`?
Can you rename it to `HnswBuilderLevelView` or `HnswLevelView`?